### PR TITLE
Handle limited open Connections due to keepalive connections (scratch pull draft pre-review, 2nd revision)

### PR DESCRIPTION
**Pull request by sgothel**
* Resolves: # 
* Target version: master 

#### Changes to original PR
- net::Config -> net::Defaults, without any xml config file side-effects
  - only allow net::Defaults to be changed in unit tests via dedicated callback
- removed Socket age criteria via Socket::creationTime and net::defaults::MaxDuration
  - Socket::creationTime still relevant for minimum throughput criteria
  - dropped the net::defaults::MaxDuration 
- max-connections
  - removed under- overflow checks
  - using SC atomic ops (not just relaxed)
- enhanced API doc of Util::TimeAverage (clarification)
- unit tests in one place / commit
- more unification/squashing to single semantic commits
- more sensible order of commits
- removed dead code
- dropped noexcept patch

Further:
- cleaned up the commit messages
- removed the COOLWSD.cpp default config reordering
- reordered the 'Unify socket stats', as it must be upfront the actual connection count limitation
  and before moved _lastSeenHTTPHeader
- fixed another intermediate breakage in the commit chain
- bisectable proof (but no unit test runs ofc) .. via
  - passed `git rebase -i --exec "make -j 20"` for all commits
   
### Checklist

- [ ] I have run `make prettier-write` and formatted the code.
- [ ] All commits have Change-Id
- [ ] I have run tests with `make check`
- [ ] I have issued `make run` and manually verified that everything looks okay
- [ ] Documentation (manuals or wiki) has been updated or is not required

### CODE REVIEW

File: net/Socket.cpp [hunk 1]
```diff
@@ -1298,18 +1495,100 @@
...
+    if (c1 || c2 || c3 || c4 )
+    {
+        LOG_WRN("CheckRemoval: Timeout: {c1 " << c1 << ", c2 " << c2 << ", c3 " << c3 << ", c4 " << c4 << "}, "
+            << getStatsString(now) << ", "
+            << *this);
+        if (_socketHandler)
+        {
+            _socketHandler->onDisconnect();
+            if( isOpen() ) {
+                // FIXME: Ensure proper semantics of onDisconnect()
```
[the remark this answers is not in the document]
Comment (sgothel) [reply net/Socket.cpp]: resolved below

File: net/WebSocketHandler.hpp [hunk 2]
```diff
@@ -599,48 +630,63 @@
...
 
 public:
+    /// Sends a native control-frame ping message
     void sendPing(std::chrono::steady_clock::time_point now,
                   const std::shared_ptr<StreamSocket>& socket)
     {
-//        assert(!_isClient);
+        if (_isClient)
+            LOG_DBG("Clients should not send pings, only servers");
+//        assert(_isClient);
```
[the remark this answers is not in the document]
Comment (sgothel) [reply net/WebSocketHandler.hpp]: resolved below

File: net/WebSocketHandler.hpp [hunk 3]
```diff
@@ -599,48 +630,63 @@
...
     }
 
+    /// Sends a native control-frame pong message
     void sendPong(std::chrono::steady_clock::time_point now,
                   const char* data, const size_t len,
                   const std::shared_ptr<StreamSocket>& socket)
     {
+        if (!_isClient)
+            LOG_WRN("Servers should not send pongs, only clients");
+//        assert(!_isClient);
```
[the remark this answers is not in the document]
Comment (sgothel) [reply net/WebSocketHandler.hpp]: resolved below

**Comment by sgothel** [comment]
now (2nd CI run) wopi-save-on-exits fails sporadic again (passed 1st time, known instable test ...)

**Comment by sgothel** [comment]
Reviewed my changes set, added missing setup in COOLWSD.cpp for maxConnection commit
> net::Defaults::get().MaxConnections = std::max<size_t>(3, MAX_CONNECTIONS);

**Comment by sgothel** [comment]
> Reviewed my changes set, added missing setup in COOLWSD.cpp for maxConnection commit
> 
> > net::Defaults::get().MaxConnections = std::max<size_t>(3, MAX_CONNECTIONS);

Added two more typos fixes: 1 comment (Socket.cpp) + 1 LOG message (WebSocketHandler ping/pong)

File: net/Socket.hpp [hunk 4]
```diff
@@ -398,12 +412,19 @@ class Socket
         , _open(_fd >= 0)
         , _creationTime(creationTime)
         , _lastSeenTime(_creationTime)
+        , _bytesSent(0)
```
Comment (caolanm) [inline net/Socket.hpp:415]: It seems that the change to move _bytesSent and _bytesRcvd into the base class is effectively only so they are available for Socket::streamStats? So I wonder if that's something we need to do?

File: wsd/COOLWSD.cpp [hunk 5]
```diff
@@ -2081,12 +2082,12 @@ void COOLWSD::innerInitialize(Application& self)
...
         { "net.connection_timeout_secs", "30" },
+        { "net.content_security_policy", "" },
+        { "net.frame_ancestors", "" },
         { "net.listen", "any" },
         { "net.proto", "all" },
-        { "net.service_root", "" },
         { "net.proxy_prefix", "false" },
-        { "net.content_security_policy", "" },
-        { "net.frame_ancestors", "" },
+        { "net.service_root", "" },
```
Comment (caolanm) [inline wsd/COOLWSD.cpp:2090]: seeing as these end up not changed, maybe drop this from this pr and put into a separate one (even though ordering them alphabetically looks a good idea)

**Review by caolanm — COMMENTED** [review]
Probably need to then update the commit messages so they just relate to the code as it is and not mention previous versions that existed during review because that context won't really apply when merged. I feel changes from earlier drafts can go in the review itself, if it's something that needs to be pointed out.

File: net/Socket.hpp (hunk 4 above)
Comment (sgothel) [reply net/Socket.hpp:415]: For stats and offering it as an criteria for checkTimeout / checkRemoval (used in the latter).
When I unified it in the Socket, idea was to push down general properties (like open) etc.

File: wsd/COOLWSD.cpp (hunk 5 above)
Comment (sgothel) [reply wsd/COOLWSD.cpp:2090]: yeah, can drop it. reordered when added the other items - now .. either way is OK.

**Comment by sgothel** [comment]
> Probably need to then update the commit messages so they just relate to the code as it is and not mention previous versions that existed during review because that context won't really apply when merged. I feel changes from earlier drafts can go in the review itself, if it's something that needs to be pointed out.

yup, will push another one tmr morning.

File: net/Socket.hpp (hunk 4 above)
Comment (sgothel) [reply net/Socket.hpp:415]: reordered the specific commit .. resulting in visible usage and compile clean commits

**Comment by sgothel** [comment]
moved to orig PR

**Pull request closed without merging**